### PR TITLE
 📖 Add tutorial for external plugin

### PR DESCRIPTION
**Pull request by Eileen-Yu**
## Description:
Add docs for the external plugin tutorial.

## Motivation:
1. Guide users to write, install, and use their own external plugins.
2. Part of the efforts of https://github.com/kubernetes-sigs/kubebuilder/discussions/3309
3. Record my experiment & tips along with my practice over the external plugin with JavaScript:
https://github.com/Eileen-Yu/kb-js-plugin

### CODE REVIEW

File: docs/book/src/plugins/external-plugins.md [hunk 1]
```diff
@@ -0,0 +1,155 @@
...
+```sh
+/path1/path2/${name}/${version} 
+```
+
+### Usage:
+
+The external plugin supports the same subcommands as kubebuilder already provides:
+- `init`: project initialization.
+- `create api`: scaffold Kubernetes API definitions.
+- `create webhook`: scaffold Kubernetes webhooks.
```
Comment (camilamacedo86) [inline docs/book/src/plugins/external-plugins.md:111]: it is missing edit

File: docs/book/src/plugins/external-plugins.md [hunk 2]
```diff
@@ -0,0 +1,155 @@
...
+
+## How to use it?
+
+### Prerequisites
+
+- You have built and installed your Kubebuilder binary.
+
+- You have built and installed your external plugin as an **executable / binary**.
+
+- Your plugin is put in a path following the GVK schema:
```
Comment (rashmigottipati) [inline docs/book/src/plugins/external-plugins.md:101]:
```suggestion
- Your plugin is placed in a path following a `group-like name and version` scheme:
```

File: docs/book/src/plugins/external-plugins.md [hunk 3]
```diff
@@ -0,0 +1,155 @@
...
+
+You can extend the Kubebuilder functionality by introducing external plugin.
+
+An external plugin is an executable that provides the same kubebuilder plugin features. Since it is an executable, such customized plugin can be implemented in any language.
+
+The Kubebuilder CLI loads the external plugin in the specified path and interact the plugin through the executable's `stdin` & `stdout`.
+
+## When is it useful?
+
+- If you want to create helpers or addons on top of the scaffolds done by the Kubebuilder internal plugins.
```
Comment (rashmigottipati) [inline docs/book/src/plugins/external-plugins.md:13]: Should we still use the terminology as "internal plugins" since all plugins will be implemented as external plugins?

File: docs/book/src/plugins/external-plugins.md [hunk 4]
```diff
@@ -0,0 +1,155 @@
+# Extending Kubebuilder with external plugins
+
+## Overview
+
+You can extend the Kubebuilder functionality by introducing external plugin.
+
+An external plugin is an executable that provides the same kubebuilder plugin features. Since it is an executable, such customized plugin can be implemented in any language.
```
Comment (rashmigottipati) [inline docs/book/src/plugins/external-plugins.md:7]: Could you elaborate a bit more on what you mean by same kubebuilder plugin features? It might be nice to include those here.

File: docs/book/src/plugins/external-plugins.md [hunk 5]
```diff
@@ -0,0 +1,155 @@
...
+
+### Prerequisites
+
+- You have built and installed your Kubebuilder binary.
+
+- You have built and installed your external plugin as an **executable / binary**.
+
+- Your plugin is put in a path following the GVK schema:
+```sh
+/path1/path2/${name}/${version}
```
[the remark this answers is not in the document]
Comment (rashmigottipati) [reply docs/book/src/plugins/external-plugins.md:103]: I like @camilamacedo86's suggestion.
> WDYT about something more like ^ ?

```suggestion
- kubebuilder CLI > 3.11.0
- Your own external plugin executable written in a language of your choice 
- Path of the external plugin configured via the ENV VAR (specify the environment variable here) or it can be placed in a path following a `group-like name and version` scheme:

```

Comment (rashmigottipati) [inline docs/book/src/plugins/external-plugins.md:103]:
```suggestion
 $HOME/.config/kubebuilder/plugins/${name}/${version} for Linux
 ~/Library/Application Support/kubebuilder/plugins/${name}/${version} for OSX
```

File: docs/book/src/plugins/external-plugins.md [hunk 6]
```diff
@@ -0,0 +1,155 @@
...
+~/Library/Application Support/kubebuilder/plugins/${name}/${version}
+```
+
+**Examples**
+```sh
+# Initialize a new project with the external plugin named `sampleplugin`
+EXTERNAL_PLUGINS_PATH=/path1/path2/sampleplugin/v1/sampleplugin kubebuilder init --plugins sampleplugin/v1
+
+# Create a new API with the above external plugin with a customized flag `number`
+EXTERNAL_PLUGINS_PATH=/path1/path2/sampleplugin/v1/sampleplugin kubebuilder create api --plugins sampleplugin/v1 --number 2
```
Comment (rashmigottipati) [inline docs/book/src/plugins/external-plugins.md:133]: Instead of `/path1/path2` can we specify something else as it may cause confusion. For example:
```suggestion
EXTERNAL_PLUGINS_PATH=<home-dir>/sampleplugin/v1/sampleplugin kubebuilder create api --plugins sampleplugin/v1 --number 2
```

This comment applies to all other places where `/path1/path2` is mentioned.

File: docs/book/src/plugins/external-plugins.md [hunk 7]
```diff
@@ -0,0 +1,155 @@
+# Extending Kubebuilder with external plugins
+
+## Overview
+
+You can extend the Kubebuilder functionality by introducing external plugin.
```
Comment (everettraven) [inline docs/book/src/plugins/external-plugins.md:5]: Nit: I would reword this a bit to be more along the lines of:
```suggestion
Kubebuilder's functionality can be extended through the use of external plugins.
```

File: docs/book/src/plugins/external-plugins.md (hunk 4 above)
Comment (everettraven) [reply docs/book/src/plugins/external-plugins.md:7]: +1 to more elaboration. I would also recommend a bit of a rephrase here to be something more along the lines of:
```suggestion
An external plugin is an executable that implements an execution pattern that Kubebuilder knows how to interact with. Since an external plugin is an executable it can be implemented in any language. 
```

File: docs/book/src/plugins/external-plugins.md (hunk 3 above)
Comment (everettraven) [reply docs/book/src/plugins/external-plugins.md:13]: +1 on avoiding "internal plugins" as terminology here. I would probably say something more along the lines of "Kubebuilder's default scaffolding"

File: docs/book/src/plugins/external-plugins.md [hunk 8]
```diff
@@ -0,0 +1,155 @@
...
+
+The Kubebuilder CLI loads the external plugin in the specified path and interact the plugin through the executable's `stdin` & `stdout`.
+
+## When is it useful?
+
+- If you want to create helpers or addons on top of the scaffolds done by the Kubebuilder internal plugins.
+
+- If you design customized layouts and want to take advantage of functions from Kubebuilder library.
+
+- If you are looking for implementing plugins by a different language other than `Go`.
```
Comment (everettraven) [inline docs/book/src/plugins/external-plugins.md:17]:
```suggestion
- If you are looking for implementing plugins in a language other than `Go`.
```

File: docs/book/src/plugins/external-plugins.md [hunk 9]
```diff
@@ -0,0 +1,155 @@
...
+	Error bool `json:"error,omitempty"`
+
+	// ErrorMsgs contains the specific error messages of the plugin failures.
+	ErrorMsgs []string `json:"errorMsgs,omitempty"`
+
+	// Flags contains the plugin specific flags that the plugin returns to Kubebuilder when it receives
+	// a request for a list of supported flags from Kubebuilder
+	Flags []Flag `json:"flags,omitempty"`
+}
+```
```
Comment (everettraven) [inline docs/book/src/plugins/external-plugins.md:79]: It might be nice to include an example of the JSON for both of these. It might be difficult for someone reading the docs to visualize the JSON based on Go type definitions.

File: docs/book/src/plugins/external-plugins.md (hunk 1 above)
Comment (everettraven) [reply docs/book/src/plugins/external-plugins.md:111]: We should also add the optional subcommands for the plugin that provides metadata to Kubebuilder

**Review by everettraven — COMMENTED** [review]
Great start @Eileen-Yu ! Here are some comments from my first review pass:

File: docs/book/src/plugins/external-plugins.md [hunk 10]
```diff
@@ -0,0 +1,155 @@
...
+
+2. On OSX
+```sh
+~/Library/Application Support/kubebuilder/plugins/${name}/${version}
+```
+
+**Examples**
+```sh
+# Initialize a new project with the external plugin named `sampleplugin`
+EXTERNAL_PLUGINS_PATH=/path1/path2/sampleplugin/v1/sampleplugin kubebuilder init --plugins sampleplugin/v1
```
[the remark this answers is not in the document]
Comment (Eileen-Yu) [reply docs/book/src/plugins/external-plugins.md:126]: Thanks for catching! 
And based on @rashmigottipati 's suggestion, hope using `export EXTERNAL_PLUGINS_PATH=<custom-path>` could be more clear 🤔

File: docs/book/src/plugins/external-plugins.md [hunk 11]
```diff
@@ -0,0 +1,168 @@
+# Extending Kubebuilder with external plugins
+
+## Overview
+
+Kubebuilder's functionality can be extended through the use of external plugins.
+
+An external plugin is an executable that implements an execution pattern that Kubebuilder knows how to interact with. Since an external plugin is an executable it can be implemented in any language. 
+
+The Kubebuilder CLI loads the external plugin in the specified path and interact the plugin through the executable's `stdin` & `stdout`.
```
Comment (everettraven) [inline docs/book/src/plugins/external-plugins.md:9]: Nit: Some suggestions for wording improvement:

```suggestion
The Kubebuilder CLI loads the external plugin in the specified path and interacts with it through `stdin` & `stdout`.
```

File: docs/book/src/plugins/external-plugins.md [hunk 12]
```diff
@@ -0,0 +1,168 @@
...
+
+- If you design customized layouts and want to take advantage of functions from Kubebuilder library.
+
+- If you are looking for implementing plugins in a language other than `Go`.
+
+## How to write it?
+
+The inter-process communication between your external plugin and Kubebuilder is through the standard I/O.
+
+You can write your external plugin in any language as long as it implements the `PluginRequest` and `PluginResponse` interfaces.
```
Comment (everettraven) [inline docs/book/src/plugins/external-plugins.md:23]: Nit: "interfaces" suggests a looser construct in programming ("interfaces" can be anything behind the scenes as long as they implement certain functionality). I would recommend using "types" as it is more strict and aligns with what the external plugin communication process in Kubebuilder expects. The response must be exactly the `PluginResponse` type and shouldn't have any additional fields

File: docs/book/src/plugins/external-plugins.md [hunk 13]
```diff
@@ -0,0 +1,168 @@
...
+
+If you need to include logs or debug messages while developing your plugin, consider writing these messages to a log file instead.
+
+</aside>
+
+## How to use it?
+
+### Prerequisites
+- kubebuilder CLI > 3.11.0
+- Your own external plugin executable written in a language of your choice
```
Comment (everettraven) [inline docs/book/src/plugins/external-plugins.md:97]: I'm not sure this should be a prerequisite - I'm interpreting this as I _have_ to write my own external plugin rather than being able to use any external plugin

File: docs/book/src/plugins/external-plugins.md [hunk 14]
```diff
@@ -0,0 +1,168 @@
...
+
+Also, there are **Optional** subcommands for a better user experience:
+- `metadata`: add customized plugin description and examples when a `--help` flag is specified. 
+- `flags`: specify valid flags for Kubebuilder to pass to the external plugin.
+
+<aside class="note">
+<h1>More about `flags` subcommand</h1>
+
+The flags subcommand in an external plugin allows for early error detection by informing Kubebuilder about the flags the plugin supports. If an unsupported flag is identified, Kubebuilder can issue an error before the plugin is called to execute. 
+If a plugin does not implement the flags subcommand, Kubebuilder will pass all flags to the plugin, which then it is the external plugin's responsibility to handle those invalid flags.
```
Comment (everettraven) [inline docs/book/src/plugins/external-plugins.md:123]: nit:
```suggestion
If a plugin does not implement the `flags` subcommand, Kubebuilder will pass all flags to the plugin, making it the external plugin's responsibility to handle any invalid flags. 
```

File: docs/book/src/plugins/external-plugins.md [hunk 15]
```diff
@@ -0,0 +1,168 @@
...
+- `edit`: update the project configuration.
+
+Also, there are **Optional** subcommands for a better user experience:
+- `metadata`: add customized plugin description and examples when a `--help` flag is specified. 
+- `flags`: specify valid flags for Kubebuilder to pass to the external plugin.
+
+<aside class="note">
+<h1>More about `flags` subcommand</h1>
+
+The flags subcommand in an external plugin allows for early error detection by informing Kubebuilder about the flags the plugin supports. If an unsupported flag is identified, Kubebuilder can issue an error before the plugin is called to execute.
```
Comment (everettraven) [inline docs/book/src/plugins/external-plugins.md:122]: Nit:
```suggestion
The `flags` subcommand in an external plugin allows for early error detection by informing Kubebuilder about the flags the plugin supports. If an unsupported flag is identified, Kubebuilder can issue an error before the plugin is called to execute. 
```

File: docs/book/src/plugins/external-plugins.md [hunk 16]
```diff
@@ -0,0 +1,168 @@
...
+- `metadata`: add customized plugin description and examples when a `--help` flag is specified. 
+- `flags`: specify valid flags for Kubebuilder to pass to the external plugin.
+
+<aside class="note">
+<h1>More about `flags` subcommand</h1>
+
+The flags subcommand in an external plugin allows for early error detection by informing Kubebuilder about the flags the plugin supports. If an unsupported flag is identified, Kubebuilder can issue an error before the plugin is called to execute. 
+If a plugin does not implement the flags subcommand, Kubebuilder will pass all flags to the plugin, which then it is the external plugin's responsibility to handle those invalid flags. 
+
+If a plugin does implement the flags subcommand, it **MUST** support `init`, `create api`, `create webhook`, and `edit` flags, corresponding to respective Kubebuilder subcommands.
```
Comment (everettraven) [inline docs/book/src/plugins/external-plugins.md:125]: nit:
```suggestion
If a plugin does implement the `flags` subcommand, it **MUST** support `--init`, `--api`, `--webhook`, and `--edit` flags, which correspond to respective Kubebuilder subcommands.
```

**Review by everettraven — COMMENTED** [review]
This doc is starting to look really good! A few more nits on my second pass:

File: docs/book/src/plugins/external-plugins.md (hunk 9 above)
Comment (Eileen-Yu) [reply docs/book/src/plugins/external-plugins.md]: Good idea!
I added sample JSONs for both request & response together with explainations.
Also, the Go definitions are now referred in the link instead of directly pasting them in the docs.

Hope this can make it clearer..

File: docs/book/src/plugins/external-plugins.md (hunk 12 above)
Comment (Eileen-Yu) [reply docs/book/src/plugins/external-plugins.md]: Thank you for the nice explanation! 👍
I tried to reorganized this paragraph in the latest commit.

File: docs/book/src/plugins/external-plugins.md (hunk 13 above)
Comment (Eileen-Yu) [reply docs/book/src/plugins/external-plugins.md]: Oh I see!
So by this line I tried to say the external plugin needs to be an executable..
I updated this part with some explanations.

File: docs/book/src/plugins/external-plugins.md [hunk 17]
```diff
@@ -0,0 +1,157 @@
...
+## How to write it?
+
+The inter-process communication between your external plugin and Kubebuilder is through the standard I/O.
+
+Your external plugin can be written in any language, given it adheres to the [PluginRequest][PluginRequest] and [PluginResponse][PluginResponse] type structures.
+
+`PluginRequest` encompasses all the data Kubebuilder collects from the CLI and previously executed plugins in the plugin chain.
+Kubebuilder conveys the marshaled PluginRequest (a `JSON` object) to the external plugin over `stdin`.
+
+Below is a sample JSON object of the `PluginRequest` type:
```
Comment (Kavinjsir) [inline docs/book/src/plugins/external-plugins.md:28]:
```suggestion
Below is a sample JSON object of the `PluginRequest` type, triggered by "kubebuilder init --plugins v1alpha1 --domain my.domain:
```

File: docs/book/src/plugins/external-plugins.md [hunk 18]
```diff
@@ -0,0 +1,157 @@
+# Extending Kubebuilder with external plugins
+
+## Overview
+
+Kubebuilder's functionality can be extended through the use of external plugins.
+
+An external plugin is an executable that implements an execution pattern that Kubebuilder knows how to interact with. Since an external plugin is an executable it can be implemented in any language.
```
Comment (Kavinjsir) [inline docs/book/src/plugins/external-plugins.md:7]:
```suggestion
An external plugin is an executable (can be implemented in any language) that implements an execution pattern that Kubebuilder knows how to interact with.
```

**Review by Kavinjsir — APPROVED** [review]
LGTM!
Left some nits which can or cannot be applied.

**Review by Kavinjsir — APPROVED** [review]

File: docs/book/src/plugins/external-plugins.md [hunk 19]
```diff
@@ -0,0 +1,157 @@
...
+`PluginResponse` represents the updated state of the project, as modified by the plugin. This data structure is serialized into `JSON` and sent back to Kubebuilder via `stdout`.
+
+Here is a sample JSON representation of the `PluginResponse` type:
+```json
+{
+    "apiVersion": "v1alpha1",
+    "command": "init",
+    "metadata": {
+        "description": "The `init` subcommand is meant to initialize a project via Kubebuilder. It scaffolds a single file: `initFile`",
+    	"Examples": "kubebuilder init --plugins sampleexternalplugin/v1 --domain my.domain"
```
Comment (everettraven) [inline docs/book/src/plugins/external-plugins.md:47]:
```suggestion
    	 "examples": "kubebuilder init --plugins sampleexternalplugin/v1 --domain my.domain"
```

File: docs/book/src/plugins/external-plugins.md (hunk 17 above)
Comment (Eileen-Yu) [reply docs/book/src/plugins/external-plugins.md:28]: Thanks for the nice suggestions👍
Just applied!

**Review by rashmigottipati — APPROVED** [review]
Nice work @Eileen-Yu, changes look good to me!

File: docs/book/src/plugins/external-plugins.md [hunk 20]
```diff
@@ -0,0 +1,157 @@
...
+## How to write it?
+
+The inter-process communication between your external plugin and Kubebuilder is through the standard I/O.
+
+Your external plugin can be written in any language, given it adheres to the [PluginRequest][PluginRequest] and [PluginResponse][PluginResponse] type structures.
+
+`PluginRequest` encompasses all the data Kubebuilder collects from the CLI and previously executed plugins in the plugin chain.
+Kubebuilder conveys the marshaled PluginRequest (a `JSON` object) to the external plugin over `stdin`.
+
+Below is a sample JSON object of the `PluginRequest` type, triggered by "kubebuilder init --plugins sampleexternalplugin/v1 --domain my.domain":
```
Comment (everettraven) [inline docs/book/src/plugins/external-plugins.md:28]: nit:
```suggestion
Below is a sample JSON object of the `PluginRequest` type, triggered by `kubebuilder init --plugins sampleexternalplugin/v1 --domain my.domain`:
```

File: docs/book/src/plugins/external-plugins.md [hunk 21]
```diff
@@ -0,0 +1,157 @@
...
+- `metadata`: add customized plugin description and examples when a `--help` flag is specified. 
+- `flags`: specify valid flags for Kubebuilder to pass to the external plugin.
+
+<aside class="note">
+<h1>More about `flags` subcommand</h1>
+
+The `flags` subcommand in an external plugin allows for early error detection by informing Kubebuilder about the flags the plugin supports. If an unsupported flag is identified, Kubebuilder can issue an error before the plugin is called to execute. 
+If a plugin does not implement the `flags` subcommand, Kubebuilder will pass all flags to the plugin, making it the external plugin's responsibility to handle any invalid flags. 
+
+If a plugin does implement the `flags` subcommand, it **MUST** support `--init`, `--api`, `--webhook`, and `--edit` flags, which correspond to respective Kubebuilder subcommands.
```
Comment (everettraven) [inline docs/book/src/plugins/external-plugins.md:111]: Nit: Now that I am thinking about it a bit more I don't think these flags are _required_ to be implemented as the external plugin would return an error about an unknown flag that Kubebuilder would ignore. That being said, we may want to consider the pros and cons of enforcing or not enforcing this behavior. 

@Eileen-Yu I'll leave it up to you if you want to change the wording to reflect that these flags aren't _required_ . IMO though this nit shouldn't block this PR from merging.

**Review by everettraven — APPROVED** [review]
This looks great @Eileen-Yu ! Thanks for putting this much needed documentation together!

I do have a couple teeny non-blocking nits:

File: docs/book/src/plugins/external-plugins.md (hunk 21 above)
Comment (Eileen-Yu) [reply docs/book/src/plugins/external-plugins.md:111]: I can understand that flag for `--init` … is more like a ‘bonus’ instead of ‘must’. For me, actually I’m still curious about the detailed scenario to support that so as to reach “fast fail”.

So I guess maybe we can drop that paragraph includes the 'must' literal.
I was imagining if it is fair enough to stay focus on the general usage(`init`, `create` and `edit`) of the external plugin.

**Comment by everettraven** [comment]
@camilamacedo86 could you give this an approval whenever you have a chance?

**Review by camilamacedo86 — APPROVED** [review]
/lgtm
